reconcile/ocm_clusters.py

```diff
@@ -1,76 +1,248 @@
+
+    if not desired_spec.server_url:
+        root_updates[ocmmod.SPEC_ATTR_SERVER_URL] = current_spec.server_url
+
+    if not desired_spec.elb_fqdn:
+        root_updates[
+            ocmmod.SPEC_ATTR_ELBFQDN
+        ] = f"elb.apps.{cluster}.{current_spec.domain}"
+
+    updates: dict[str, Any] = {}
+    updates[ocmmod.SPEC_ATTR_PATH] = desired_spec.path
+    updates["root"] = root_updates
+    updates["spec"] = ocm_spec_updates
+
+    return updates, error
+
+
+def get_cluster_ocm_update_spec(
+    ocm: ocmmod.OCM, cluster: str, current_spec: OCMSpec, desired_spec: OCMSpec
+) -> Tuple[dict[str, Any], bool]:
+    """Get cluster updates to request to OCM api
+
+    Parameters:
+        ocm (OCM): ocm implementation for an ocm product (osd, rosa)
+        cluster (str): cluster name
+        current_spec (OCMSpec): Cluster spec retreived from OCM api
+        desired_spec (OCMSpec): Cluster spec retreived from App-Interface
+
+    Returns:
+        updates (dict): Updates to request to OCM api
+        error (bool): If errors detected due to spec not allowed updates.
+    """
+
+    impl = ocmmod.OCM_PRODUCTS_IMPL[current_spec.spec.product]
 
     error = False
-    if current_spec["network"] != desired_spec["network"]:
+    if not desired_spec.network.type:
+        desired_spec.network.type = "OpenShiftSDN"
+
+    if current_spec.network != desired_spec.network:
         error = True
-        logging.error(f"[{cluster_name}] invalid update: network")
-    current_spec_spec = current_spec["spec"]
-    desired_spec_spec = desired_spec["spec"]
-    updated = {
-        k: desired_spec_spec[k]
-        for k in desired_spec_spec
-        if current_spec_spec.get(k) != desired_spec_spec[k]
+        logging.error(f"[{cluster}] invalid update: network")
+
+    # Convert ocm specs to dicts, removing null values and excluded attributes
+    current_ocm_spec = {
+        k: v
+        for k, v in current_spec.spec.dict().items()
+        if v is not None and k not in impl.EXCLUDED_SPEC_FIELDS
     }
 
-    # we only need deleted to check if a field removal is valid
-    # we really want to check updated + deleted, and since
-    # we have no further use for deleted -
-    deleted = {
-        k: current_spec_spec[k] for k in current_spec_spec if k not in desired_spec_spec
+    desired_ocm_spec = {
+        k: v
+        for k, v in desired_spec.spec.dict().items()
+        if v is not None and k not in impl.EXCLUDED_SPEC_FIELDS
     }
-    diffs = deleted
-    diffs.update(updated)
 
-    invalid_fields = set(diffs.keys()) - ALLOWED_SPEC_UPDATE_FIELDS
-    if invalid_fields:
+    # Updated attributes in app-interface
+    updated_attrs = {
+        k: v for k, v in desired_ocm_spec.items() if current_ocm_spec.get(k) != v
+    }
+
+    # Removed attributes in app-interface
+    deleted_attrs = {
+        k: v
+        for k, v in current_ocm_spec.items()
+        if k not in desired_ocm_spec and v is not None
+    }
+
+    diffs = deleted_attrs
```

Comment:
It doesn't look like you end up using `deleted_attrs` again, but it's worth noting this behavior (that you're probably already aware of):

```python
>>> deleted_attrs = {'deleted': 'attrs'}
>>> diffs = deleted_attrs
>>> diffs.update({'updated': 'attrs'})
>>> deleted_attrs
{'deleted': 'attrs', 'updated': 'attrs'}
>>> diffs
{'deleted': 'attrs', 'updated': 'attrs'}
```

Normally using `copy.deepcopy()` could help when we want to ensure that we're not using a reference to the existing object, but instead an independent copy, but maybe `diffs = deleted_attrs | updated_attrs` could lead to less unexpected behavior later on (if say we added logging for these vars).

Reply:
I just want to be clear on semantics. As you have already seen, this dict is not used further, and IMO is better to read it as `diffs` instead of using `deleted_attrs` or `updated_attrs` as the base to compare with the `ALLOWED_SPEC_UPDATE_FIELDS` constant.

Reply:
The variable name `diffs` makes sense to me, but really what I'm suggesting is what you're doing is assigning two variable names that reference the same object in memory, yet they have different meanings. `deleted_attrs` makes sense for describing what that object in memory is, until you also create the `diffs` reference and then start changing the underlying data. That's why I was suggesting that `diffs = deleted_attrs | updated_attrs` makes sense because it creates a new object in memory, such that `diffs` and `deleted_attrs` now reference to separate objects.

For now, it's not an issue the way the code is working today, but it could quite easily become problematic if someone changes the code by logging `deleted_attrs` or using it in some other way. This may seem like a nit, but I've seen this issue plenty of times when people who change code later either aren't familiar with how Python variables are references to objects, or they haven't noticed that the same object was referenced with multiple variable names.

Reply:
Fair enough. I changed it with your suggestion.